### PR TITLE
feat(avalonia): implement World Map Image (FE7) editor (#1184)

### LINKED ISSUE #1184: Avalonia: implement World Map Image (FE7) editor (port WinForms WorldMapImageFE7Form)
FE7 variant of the world-map image editor (326 lines) with image preview + import/export; the base FE8 view is real but the FE7 class is a separate stub. The Avalonia WorldMapImageFE7View is a 21-line scaffold with no image surface. Gap type: stub. Acceptance: Implement a functional Avalonia World Map Image (FE7) editor with parity to WinForms WorldMapImageFE7Form.

## Category
Misc editors

## Acceptance
Implement a functional Avalonia editor/tool with parity to the referenced WinForms form (list + edit controls + write-back/preview as applicable), wired into the Avalonia editor registry, with unit/parity tests and a real-GUI screenshot.

---
🔎 Filed by [Claude Code](https://claude.com/claude-code) `v2.1.169` · model `claude-opus-4-8[1m]` — part of the **Avalonia GUI stub audit (wave 2)**: a verified WinForms↔Avalonia menu-form diff found 36 user-facing editors/tools that exist only as bare Avalonia scaffolds/stubs and were not covered by the first audit (#994–#1037).

### PR DESCRIPTION
**Pull request by laqieer**
## Summary

Ports WinForms `WorldMapImageFE7Form` (326 lines) to a functional Avalonia **World Map Image (FE7)** editor. The previous `WorldMapImageFE7View` was a 21-line scaffold; it now renders the FE7 12-split big field map + the event image, with FE7 import / export / pointer write-back. Plan + review on issue #1184.

## What changed

**Two new pure-Core helpers** (`FEBuilderGBA.Core/ImageWorldMapCore.cs`):
- **`TryRenderFE7BigFieldMap(rom)`** (READ-ONLY) — ports `ImageUtilMap.DrawWorldMapFE7`: resolves the shared 4-bank palette + the two 12-entry pointer arrays (`worldmap_big_image_pointer` → images, `worldmap_big_palettemap_pointer` → the "TSA 12-split" header-TSAs), decodes each 256×256 header-TSA piece via `ImageUtilCore.DecodeHeaderTSA`, and composites the 12 pieces into one 1024×688 RGBA canvas (bottom row clipped to 176, but TSA still decoded at 256 — the WF quirk). **FE7-only** (explicit `version == 7` gate). The FE7 big-map image + TSA are RAW (uncompressed), unlike the event image (LZ77) — corrects the LZ77 assumption.
- **`ImportFE7BigFieldMap(rom, rgba, w, h)`** (ROM-MUTATING) — the inverse of the render: validate-all-before-mutate (pre-resolves + bounds-checks all 25 destinations), derives a 4-bank palette + per-tile banks, encodes each piece (plain non-dedup TSA + `EncodeHeaderTSA(margin:0)`, bottom-row image sliced to `256/2*176`), and writes all 12 image + 12 TSA + palette regions **RAW in-place** (fixed size, no repoint) under the caller's **ambient undo only** (no `Undo.UndoData` param), with a defensive byte-identical fault restore.
- **`CanImportFE7BigFieldMap(rom)`** (READ-ONLY) — lightweight version/pointer gate for the Import button (no render allocation).

**Event preview/import reuse**: the existing version-agnostic `TryRenderEvent` + `ImportEvent`. The FE8-only `ImportEvent` gate (#1098) is generalized to FE7+FE8 — the world-map event format is identical and the encode path is version-independent; FE6 (pointers 0x0) is still rejected.

**ViewModel + View + AXAML**: two `GbaImageControl` previews (big map + event) with per-section Import (PNG dialog → remap → Core import under one `UndoService` scope) and Export, a Write-Pointers button (WF `WriteButton_Click` + `WMEvent.WritePointer` parity), an Undo button, and three read-only pointer labels. All handlers use `Log.Error("...: " + ex.ToString())` (Core `Log.Error` is `params string[]`, not composite-format). `SelectFirstItem()` triggers the render for `--screenshot-all`.

**Localization**: ja/zh entries for the new XAML literals (`Big Field Map`, `Event Image`, `Image/Palette/TSA Pointer`).

## Addresses all 4 Copilot plan-review findings
1. Explicit FE7 `version == 7` gate on both render + import (FE8 → null / error, zero mutation).
2. Event-import parity — generalized the `ImportEvent` gate to FE7 + wired the Write-Pointers button.
3. No double-undo — the Core import takes no `Undo.UndoData`; the View owns the single ambient `UndoService` scope.
4. Validate-all-before-mutate — all 25 destinations resolved + bounds-checked before the first write; exact write sizes asserted.

## Scope
- Closes #1184.

## Test plan

- [x] `dotnet build` Core + Avalonia + SkiaSharp — 0 errors.
- [x] `ImageWorldMapFE7CoreTests` (11 tests) — real FE7U render returns non-null 1024×688; render null on null-ROM and FE8; `CanImport` true/false on FE7/FE8; import round-trip (render→export RGBA→reimport) returns "" success and mutates ROM; FE8 / wrong-dims / null-image imports return a non-empty error with **zero** ROM mutation (snapshot before/after).
- [x] `WorldMapImageFE7ViewModelTests` (3 tests) — preview delegates forward the Core result faithfully with FE7U; `CanImport` true on FE7 / false on FE8; no-ROM guard returns null/false.
- [x] `ImageWorldMapEventImportTests` updated — FE7 now passes the version gate (new test); FE6 rejected with zero mutation (replaces the old FE7-rejection test); all 17 event-import tests green.
- [x] All `WorldMapImage*` Avalonia tests green together (65 passed) — no regression in the base view.
- [x] FEBuilderGBA.Tests `AutomationId | AvaloniaViews | AvaloniaFieldCompleteness` filter — 26 passed.
- [x] `scripts/validate-automation-ids.ps1` — VALIDATION PASSED (0 naming errors, 0 duplicates; new IDs end in `_Label/_Button/_Image`).
- [x] Screenshot proof: `--screenshot-all` on FE7U.gba rendered a **non-blank** FE7 Elibe world map (605 KB PNG).

> Note: a full local `Core.Tests` run shows ~110 failures in `StructExportCoreTests` / `ExportFilterCoreTests` / `SkillSystemsAnime*` — these are a **pre-existing cross-class shared-state ordering artifact** (they pass in isolation and fail even with my test class absent; unrelated to these World-Map files). CI runs with proper isolation.

## GUI Test Report

| Area | Test | Result |
|------|------|--------|
| Big field map render (FE7U) | `--screenshot-all` renders the 1024×688 Elibe map non-blank | PASS |
| Pointer labels | Image / Palette / TSA pointer values shown | PASS |
| Buttons present | Write Pointers / Undo / Import / Export | PASS |
| Event preview | reuses version-agnostic `TryRenderEvent` | PASS |
| Import gate | `CanImport` enables Import only on FE7 | PASS (VM test) |

Fixes #1184

Generated with Claude Code (Opus 4.8 1M context)
Model: claude-opus-4-8[1m]

### CODE REVIEW

**Review by laqieer — COMMENTED** [review]
Metadata gates passed: the screenshot is a GitHub release asset and shows the actual Avalonia World Map Image (FE7) GUI with populated pointer values and the map preview; the PR has a `## GUI Test Report` table with PASS entries; and every `## Test plan` item is checked.

Requesting changes for these additional blocking issues:

1. `FEBuilderGBA.Core/ImageWorldMapCore.cs:434` - `TryRenderFE7BigFieldMap` renders each FE7 piece through `ImageUtilCore.DecodeHeaderTSA`, but that Core renderer skips `tsatile == 0` as blank. The WinForms path this ports (`ImageUtil.ByteToImage16TileInner`) only skips `0xFFFF`, and this PR's own `EncodeFE7PieceTSA` emits `entry = n & 0x3FF`, so `n == 0` is a valid tile-0 reference. As written, the first tile in every 256x256 piece is blank in preview/export, and a render->import round trip can bake those blanks back into the ROM. Please use or add a FE7 header-TSA decode path that treats tile 0 as valid, and add a regression test with a non-empty tile 0.

2. `FEBuilderGBA.Core/ImageWorldMapCore.cs:566` - `ImportFE7BigFieldMap` builds a 4-bank palette with real source colors starting at slot 0, then immediately remaps through `ImageImportCore.RemapToMultiPalette`, which skips palette index 0 for all opaque pixels. WinForms `ImageToByte16Tile` preserves the source low nibble, including index 0, and the FE7 big map is an opaque background where slot 0 is usable. This silently reduces each bank to 15 usable colors and corrupts imports that need or select slot 0. Either reserve slot 0 consistently and reject/handle the 15-color limit, or use a FE7 big-map remapper that can assign opaque pixels to index 0.

3. `FEBuilderGBA.Avalonia/Views/WorldMapImageFE7View.axaml.cs:200` / `:249` - the Event Export button saves the full rendered event preview (256x160), but `DoEventImport` rejects anything except 240x160. A user cannot export the event image from this editor and import it back without manually cropping the 16px margin. Please crop exports to the visible 240x160 source size, or accept 256x160 imports and strip the right margin before calling `ImportEvent`.

The existing unresolved review threads also look valid and should be addressed: `WritePointers()` currently creates a dirty undo entry while writing each slot to its existing value, and `EventImportButton` is gated by the big-map `CanImport` flag instead of the event pointer prerequisites.

Copilot CLI: 1.0.64-0
Model: GPT-5.5 (gpt-5.5)

**Comment by laqieer** [comment]
All 5 review findings fixed in 528fc252e (2 inline bot threads + 3 CLI blocking issues):

**CLI #1 (blocking) — tile-0 rendered blank.** `ImageUtilCore.DecodeHeaderTSA` skipped a TSA cell of `0x0000` as blank (the sprite/CG default), but WF `ByteToImage16TileInner` only skips `0xFFFF`. Added a `skipTile0` parameter (default `true`, preserving BigCG/OP-Prologue byte-for-byte — existing `ImageUtilCoreHeaderTSATests` stay green); `TryRenderFE7BigFieldMap` now passes `skipTile0:false` so tile 0 is a valid background tile. New regression test `DecodeHeaderTSA_SkipTile0False_RendersTile0` proves a tile-0 cell renders its color (opaque) instead of transparent.

**CLI #2 (blocking) — index-0 palette loss.** `RemapToMultiPalette` reserves index 0 for transparency (15 usable colors/bank), but the FE7 big map is an opaque background where slot 0 is a real color. Added `RemapFE7BigField` which maps every pixel to the closest of **all 16** colors (index 0 included) in its tile's chosen bank.

**CLI #3 (blocking) — event export/import dim mismatch.** Event Export saved the 256×160 preview but `DoEventImport` accepts only 240×160. Event Export now crops to the visible 240×160 (`CropTopLeft`), so an export→import round trip succeeds without manual cropping.

**Bot thread (Event Import gate).** Added `CanImportEvent` (FE7+FE8, gated on the `worldmap_event_*` pointers — independent of the big-map pointers) and bound `EventImportButton` to it. Core + VM tests.

**Bot thread (`WritePointers` no-op dirty state).** The pointers are read-only labels and the in-place big-map import never moves data, so "Write Pointers" only created a spurious dirty/undo entry. Removed the button + `WritePointers()`; the top-level Undo button reverts the most recent import.

Tests: 49 Core (FE7 + HeaderTSA + event-import) + 3 Avalonia VM, all green. Build Core/Avalonia/SkiaSharp 0 errors. `validate-automation-ids.ps1` PASSED. Gate-4 filter 26/26. Screenshot re-captured (still a non-blank FE7 Elibe map) and re-uploaded to the release asset.

---
Generated with Claude Code (Opus 4.8 1M context) — model `claude-opus-4-8[1m]`

File: FEBuilderGBA.Avalonia/ViewModels/WorldMapImageFE7ViewModel.cs [hunk 1]
```diff
@@ -24,10 +57,78 @@ public List<AddrResult> LoadList()
...
+                rom.u32(rom.RomInfo.worldmap_big_palette_pointer));
+            rom.write_u32(rom.RomInfo.worldmap_big_palettemap_pointer,
+                rom.u32(rom.RomInfo.worldmap_big_palettemap_pointer));
+            rom.write_u32(rom.RomInfo.worldmap_event_image_pointer,
+                rom.u32(rom.RomInfo.worldmap_event_image_pointer));
+            rom.write_u32(rom.RomInfo.worldmap_event_palette_pointer,
+                rom.u32(rom.RomInfo.worldmap_event_palette_pointer));
+            rom.write_u32(rom.RomInfo.worldmap_event_tsa_pointer,
+                rom.u32(rom.RomInfo.worldmap_event_tsa_pointer));
+        }
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/ViewModels/WorldMapImageFE7ViewModel.cs]: Fixed in 528fc252e: the 'Write Pointers' button + WritePointers() are removed. The pointers are read-only display labels and the in-place big-map import never moves data, so this button only created the spurious dirty/undo entry you flagged. The top-level Undo button reverts the most recent import.

File: FEBuilderGBA.Avalonia/Views/WorldMapImageFE7View.axaml.cs [hunk 2]
```diff
@@ -9,49 +19,286 @@ namespace FEBuilderGBA.Avalonia.Views
...
+        }
+
+        void UpdateUI()
+        {
+            ImagePtrLabel.Text = string.Format("0x{0:X08}", _vm.BigImagePointer);
+            PalettePtrLabel.Text = string.Format("0x{0:X08}", _vm.BigPalettePointer);
+            TsaPtrLabel.Text = string.Format("0x{0:X08}", _vm.BigTsaPointer);
+            BigImportButton.IsEnabled = _vm.CanImport;
+            EventImportButton.IsEnabled = _vm.CanImport;
+        }
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/Views/WorldMapImageFE7View.axaml.cs:54]: Fixed in 528fc252e: added ImageWorldMapCore.CanImportEvent (FE7+FE8, gated on the three worldmap_event_* pointers) and bound EventImportButton to _vm.CanImportEvent instead of the big-map _vm.CanImport. Big-map Import stays on CanImport.

**Review by laqieer — COMMENTED** [review]
No blocking findings in the re-review of 528fc252e.

Verified the five prior findings are resolved: `DecodeHeaderTSA(skipTile0:false)` is used for the FE7 big-map render and has a tile-0 regression test; FE7 big-map import now uses `RemapFE7BigField` and includes index 0; event export crops to the 240x160 import size; Event Import is gated by `CanImportEvent` on the event pointers; and the no-op Write Pointers button/method was removed.

Metadata gates pass: the PR title starts with `feat`, the description contains a release-asset `` that renders the actual Avalonia World Map Image (FE7) GUI with populated pointer values and map data, there are no feature-branch blob/raw screenshot URLs, `## GUI Test Report` includes a pass/fail table, and every `## Test plan` item is checked `[x]`. The targeted Core and Avalonia test slices for the changed world-map code passed locally.

Copilot CLI: 1.0.64-0
Model: GPT-5.5 (gpt-5.5)

File: FEBuilderGBA.Avalonia/Views/WorldMapImageFE7View.axaml.cs [hunk 3]
```diff
@@ -9,49 +19,317 @@ namespace FEBuilderGBA.Avalonia.Views
...
+
+        void RenderInto(GbaImageControl target, Func<IImage?> render, string label)
         {
             try
             {
-                _vm.LoadEntry(addr);
-                UpdateUI();
+                IImage? img = render();
+                target.SetImage(img);
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/Views/WorldMapImageFE7View.axaml.cs]: Fixed in 7b1fe84b3: RenderInto now does `using IImage? img = render();` so the Core-rendered image is disposed after SetImage copies its pixels into the GbaImageControl bitmap.

File: FEBuilderGBA.Core/ImageWorldMapCore.cs [hunk 4]
```diff
@@ -286,6 +287,605 @@ public static IImage TryRenderMainFieldMap(ROM rom)
...
+                {
+                    int dx = dstX + x;
+                    if (dx < 0 || dx >= dstW) continue;
+                    int si = (y * srcW + x) * 4;
+                    int di = (dy * dstW + dx) * 4;
+                    if (si + 3 >= s.Length || di + 3 >= dst.Length) continue;
+                    dst[di + 0] = s[si + 0];
+                    dst[di + 1] = s[si + 1];
+                    dst[di + 2] = s[si + 2];
+                    dst[di + 3] = 255; // opaque
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Core/ImageWorldMapCore.cs]: Fixed in 7b1fe84b3: BlitPiece now copies the SOURCE alpha (`dst[di+3] = s[si+3]`) instead of forcing 255. Visually identical for the opaque FE7 big map (re-captured screenshot is byte-identical), but no longer turns alpha-0 cells into opaque black in any shared use.

File: FEBuilderGBA.Core.Tests/ImageWorldMapFE7CoreTests.cs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            {
+                byte[] before = (byte[])rom.Data.Clone();
+                string err = ImageWorldMapCore.ImportFE7BigFieldMap(rom, null, FE7_BIG_WIDTH, FE7_BIG_HEIGHT);
+                Assert.False(string.IsNullOrEmpty(err));
+                Assert.Equal(before, rom.Data);
+            });
+        }
+
+        [Fact]
+        public void Import_RoundTrip_FE7_Succeeds_AndMutates()
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Core.Tests/ImageWorldMapFE7CoreTests.cs]: Fixed in 7b1fe84b3: renamed to Import_RoundTrip_FE7_Succeeds_WritesInPlaceUnderUndo and added real post-conditions — palette slot NOT repointed (in-place), the 128-byte palette region was actually re-written, the change is recorded under the ambient undo (undoData.list non-empty), and undo restores the ROM byte-identically.

File: FEBuilderGBA.Avalonia/ViewModels/WorldMapImageFE7ViewModel.cs [hunk 6]
```diff
@@ -1,16 +1,55 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// World Map Image (FE7) editor ViewModel (#1184) — a port of WinForms
+// WorldMapImageFE7Form. Exposes two read-only IImage previews (the 12-split
+// big field map + the event image) and the FE7-only import / pointer write-back,
+// all delegating to the pure ImageWorldMapCore helpers. Mirrors the preview /
+// import / pointer idiom of the base WorldMapImageViewModel.
```
[the remark this answers is not in the document]
Comment (laqieer) [reply FEBuilderGBA.Avalonia/ViewModels/WorldMapImageFE7ViewModel.cs]: Fixed in 7b1fe84b3: the VM header comment now describes the read-only display labels + in-place import and explicitly notes there is NO pointer write-back (the WF Write Pointers button is omitted).
